File: client/src/Vessel.js
          function Vessel(jsonObject){
            this.mmsi = jsonObject.userid? jsonObject.userid:jsonObject.mmsi; //notwendig, weil posEvents userid statt mmsi haben
            this.msgid = jsonObject.msgid;
            this.name = jsonObject.name;
            this.time_received = jsonObject.time_received;
            this.cog = jsonObject.cog;
            this.sog = jsonObject.sog;
            this.lat = jsonObject.pos[1]            
            this.lon = jsonObject.pos[0];
            this.imo = jsonObject.imo;
            this.true_heading = jsonObject.true_heading;
            this.dim_port = jsonObject.dim_port;
            this.dim_stern = jsonObject.dim_stern;
            this.dim_bow = jsonObject.dim_bow;
            this.dim_starboard = jsonObject.dim_starboard;
            this.ship_type = jsonObject.ship_type;
            this.nav_status = jsonObject.nav_status;
            this.dest = jsonObject.dest;
            this.draught = jsonObject.draught;

            
            this.updatePosition = function(jsonObject){
              this.lat = jsonObject.pos[1]            
              this.lon = jsonObject.pos[0];
              this.msgid = jsonObject.msgid;
              this.time_received = jsonObject.time_received;
              this.cog = jsonObject.cog;
              this.sog = jsonObject.sog;
              this.true_heading = jsonObject.true_heading;
            }

            this.paintToMap = function(zoom, callback){
              if(this.lat != null)
              {    
                var moving = (this.sog && this.sog > 0.4 && this.sog!=102.3) ; //nur Schiffe, die sich mit mind. 0,3 Knoten bewegen
                var shipStatics = (this.cog ||(this.true_heading &&  this.true_heading!=0.0 &&  this.true_heading !=511)) && (this.dim_port && this.dim_stern) ;
         
                var angle = calcAngle(this);
                var cos_angle=Math.cos(angle);
                var sin_angle=Math.sin(angle);
                var vectorPoints = [];
                var shipPoint = new L.LatLng(this.lat,this.lon);
                vectorPoints.push(shipPoint);
                if (moving) // zeichne für fahrende Schiffe einen Speedvector, ein Richtungsdreieck und möglichst ein Polygon
                {
                  vectorPoints.push(shipPoint);
                  vectorPoints.push(shipPoint);
                  var vectorLength = this.sog >30? this.sog/10: this.sog;
                  var targetPoint = calcVector(this.lon,this.lat, vectorLength, sin_angle, cos_angle);
                  vectorPoints.push(targetPoint);
                  var vectorWidth = (this.sog > 30?5:2); 
                  this.vector = L.polyline(vectorPoints, {color: 'red', weight: vectorWidth });
                
                  if (shipStatics)
                  {
                    this.polygon = new L.animatedPolygon(vectorPoints,{
                                                           autoStart:false,
                                                           distance: vectorLength/10,
                                                           interval: 200,
                                                           dim_stern: this.dim_stern,
                                                           dim_port: this.dim_port,
                                                           dim_bow: this.dim_bow,
                                                           dim_starboard: this.dim_starboard,
                                                           angle: angle,
                                                           color: "blue",
                                                           weight: 3,
                                                           fill:true,
                                                           fillColor:shipTypeColors[this.ship_type],
                                                           fillOpacity:0.6,
                                                           clickable:false,
                                                           animation:true
                    });
                  }

                  this.feature = L.animatedPolygon(vectorPoints,{
                                                          autoStart: false,
                                                          distance: vectorLength/10,
                                                          interval:200,
                                                          angle: angle,
                                                          zoom: zoom,
                                                          color: "black",
                                                          weight: 1,
                                                          fill:true,
                                                          fillColor:shipTypeColors[this.ship_type],
                                                          fillOpacity:0.8,
                                                          clickable:true,
                                                          animation:true
                  })
                }
                else //zeichne für nicht fahrende Schiffe einen Circlemarker und möglichst ein Polygon
                {
                  if(shipStatics)
                  {
                    this.polygon = L.animatedPolygon( vectorPoints,{
                                                           dim_stern: this.dim_stern,
                                                           dim_port: this.dim_port,
                                                           dim_bow: this.dim_bow,
                                                           dim_starboard: this.dim_starboard,
                                                           angle: angle,
                                                           color: "blue",
                                                           weight: 3,
                                                           fill:true,
                                                           fillColor:shipTypeColors[this.ship_type],
                                                           fillOpacity:0.6,
                                                           clickable:false,
                                                           animation:false
                    });
                  }
                  var circleOptions = {
                              radius:4,
                              fill:true,
                              fillColor:shipTypeColors[this.ship_type],
                              fillOpacity:0.8,
                              color:"#000000",
                              strokeOpacity:1,
                              strokeWidth:0.5
                  };
                   this.feature = L.circleMarker(vectorPoints[0], circleOptions);
                }
              }
          callback([this.vector,this.polygon, this.feature], getPopupContent(this));
        };

        function getPopupContent(vessel){
              var timeNow = new Date();
              var mouseOverPopup ="<div><table>";
              if(vessel.name) mouseOverPopup+="<tr><td colspan='2'><b>"+vessel.name+"</b></nobr></td></tr>";
              if(vessel.imo)mouseOverPopup+="<tr><td>IMO</td><td>"+(vessel.imo)+"</b></nobr></td></tr>  ";
              mouseOverPopup+="<tr><td>MMSI: &nbsp;</td><td><nobr>"+(vessel.mmsi)+"</nobr></td></tr>";
              if(vessel.nav_status && vessel.nav_status < 15 && vessel.nav_status > -1)
              {
                mouseOverPopup+="<tr><td>NavStatus: &nbsp;</td><td><nobr>"+ nav_stati[( vessel.nav_status)]+"</nobr></td></tr>";
              }
              if( vessel.sog)mouseOverPopup+="<tr><td>Speed: &nbsp;</td><td><nobr>"+( vessel.sog)+"</nobr></td></tr>";
              if( vessel.true_heading &&  vessel.true_heading != 511)
              {
                 mouseOverPopup+="<tr><td>Heading: &nbsp;</td><td><nobr>"+(vessel.true_heading)+"</nobr></td></tr>";
              }
              if(vessel.cog)mouseOverPopup+="<tr><td>Course: &nbsp;</td><td><nobr>"+(vessel.cog)+"</nobr></td></tr>";
             
              mouseOverPopup+="<tr><td>TimeReceived: &nbsp;</td><td><nobr>"+createDate(vessel.time_received)+"</nobr></td></tr>";
              if(vessel.dest) mouseOverPopup+="<tr><td>Dest</td><td>"+(vessel.dest)+"</b></nobr></td></tr>";
              if(vessel.draught) mouseOverPopup+="<tr><td>draught</td><td>"+(vessel.draught/10)+"</b></nobr></td></tr>";
              if(vessel.dim_bow && vessel.dim_port)mouseOverPopup+="<tr><td>width, length</td><td>"+(vessel.dim_starboard + vessel.dim_port)+", "+(vessel.dim_stern + vessel.dim_bow )+"</b></nobr></td></tr>";
              if(shipTypes[(vessel.ship_type)]) mouseOverPopup+="<tr><td>ship_type</td><td>"+ shipTypes[(vessel.ship_type)]+"</b></nobr></td></tr>";
              mouseOverPopup+="</table></div>";
              return mouseOverPopup;
            }
      function calcVector(lon, lat, sog, sin, cos){
      var dy_deg = -(sog * cos)/10000;
      var dx_deg = -(- sog * sin)/Math.cos((lat)*(Math.PI/180.0))/10000;
      return new L.LatLng(lat - dy_deg, lon - dx_deg);
    }

    function calcPoint(lon, lat, dx, dy, sin_angle, cos_angle){
      var dy_deg = -((dx*sin_angle + dy*cos_angle)/(1852.0))/60.0;
      var dx_deg = -(((dx*cos_angle - dy*sin_angle)/(1852.0))/60.0)/Math.cos(lat * (Math.PI /180.0));
      return new L.LatLng(lat - dy_deg, lon - dx_deg);
    }



function calcAngle (vessel) {
       //benötigte Daten
       var hdg = vessel.true_heading;
       var cog = vessel.cog;
       var sog = vessel.sog;
       var direction = 0;
       if (vessel.mmsi == 211855000)
       {
        direction = 299;
       }
       if (sog && sog > 0.4 && cog < 360)
       {
          direction = cog;
       }
       else if  ( hdg >0.0 && hdg !=511 && hdg < 360)
       {
         direction = hdg;
       }
       return (-direction *(Math.PI / 180.0));
   }

   

    function createDate(ts, sec, msec){
      var returnString;
      var date= new Date();
          date.setTime(ts);

      var month = date.getMonth()+1;
      var day = date.getDate();
      returnString = day +"."+month+" ";

      var hour = date.getHours();
      var min= date.getMinutes();
      returnString += addDigi(hour)+":"+addDigi(min);
      if (sec)
      {
        var seconds = date.getSeconds();
        returnString += " "+addDigi(seconds);
      }
      if (msec)
      {
        var milliseconds = date.getMilliseconds();
        returnString += " "+addDigi(milliseconds);
      }
      return returnString;
    }

    function addDigi(curr_min){
    curr_min = curr_min + "";
      if (curr_min.length == 1)
      {
        curr_min = "0" + curr_min;
      }
      return curr_min;
    }

    var shipTypes = {
                  6:'Passenger Ships',
                  7: 'Cargo Ships',
                  8: 'Tankers',
                  30:'Fishing',
                  31:'Towing',
                  32:'Towing',
                  33:'Dredger',
                  34:'Engaged in diving operations',
                  35:'Engaged in military operations',
                  36: 'Sailing',
                  37: 'Pleasure craft',
                  50:'Pilot vessel',
                  51:'Search and rescue vessels',
                  52:'Tugs',53:'Port tenders',
                  54:'anti-pollution vessels',
                  55:'Law enforcement vessels',
                  56:'Spare for local vessels',
                  57:'Spare for local vessels',
                  58:'Medical transports',
                  59:'Ships according to RR'
                };

var shipTypeColors = {
                  20:'#f9f9f9',
                  29:'#f9f9f9',
                  30:'#f99d7b'/*brown, Fishing*/,
                  31:'#4dfffe'/*lightblue, Towing*/,
                  32:'#4dfffe'/*lightblue, Towing*/,
                  33:'#f9f9f9'/*gray, Dredger*/,
                  34:'white'/*Engaged in diving operations*/,
                  35:'white'/*Engaged in military operations*/,
                  36:'#f900fe'/*violett, Sailing*/,
                  37:'#f900fe'/*violett, Pleasure craft*/,
                  40:'#f9f9f9'/*Highspeed*/,
                  49:'#f9f9f9'/*Highspeed*/,  
                  50:'red'/*Pilot vessel*/,
                  51:'white' /*Search and rescue vessels*/,
                  52:'#4dfffe'/*lightblue, Tugs*/,
                  53:'#4dfffe'/*lightblue, Port tenders*/,
                  54:'white'/*anti-pollution vessels*/,
                  55:'white'/*Law enforcement vessels*/,
                  56:'#d2d2d2'/*not classified => used as default by vesseltracker*/,
                  57:'white'/*Spare for local vessels*/,
                  58:'white'/*Medical transports*/,
                  59:'white'/*Ships according to RR*/,
                  6:'#2d00fe'/*blue, Passenger Ships*/,
                  60:'#2d00fe'/*blue, Passenger Ships*/,
                  69:'#2d00fe'/*blue, Passenger Ships*/,
                  7: '#95f190'/*lightgreen, Cargo Ships*/,
                  70:'#95f190'/*lightgreen, Cargo Ships*/,
                  79:'#95f190'/*lightgreen, Cargo Ships*/,
                  8: '#f70016'/*red, Tankers*/,
                  80:'#f70016'/*Tanker*/,
                  89:'#f70016'/*red,Tankers*/,
                  9:'#d2d2d2'/*Other Type*/,
                  90:'#d2d2d2'/*Other Type*/,
                  99:'#d2d2d2'/*Other Type*/
}
var nav_stati = {
                  0:'under way us. engine',
                  1:'at anchor',
                  2: 'not under command',
                  3: 'restr. maneuverability',
                  4: 'constr. by draught',
                  5: 'moored',
                  6: 'aground',
                  7: 'engaged in fishing',
                  8: 'under way sailing',
                  9: 'future use',
                  10: 'future use',
                  11: 'future use',
                  12: 'future use',
                  13: 'future use',
                  14: 'AIS-SART (active)',
                  15: 'not defined' 
                }

var aton_types = {
                  0:'notSpecified',
                  1:'ReferencePoint',
                  2: 'RACON',
                  3: 'off-shoreStructure',
                  4: 'futureUse',
                  5: 'LightWithoutSectors',
                  6: 'LightWithSectors',
                  7: 'LeadingLightFront',
                  8: 'LeadingLightRear',
                  9: 'BeaconCardinalN',
                  10: 'BeaconCardinalE',
                  11: 'BeaconCardinalS',
                  12: 'BeaconCardinalW',
                  13: 'BeaconPorthand', 
                  14: 'BeaconStarboardhand',
                  15: 'BeaconPreferredChannelPortHand',
                  16: 'BeaconPreferredChannelStarboardHand',
                  17: 'BeaconIsolatedDanger',
                  18: 'BeacoSafeWater',
                  19: 'BeaconSpecialMark',
                  20: 'CardinalMarkN',
                  21: 'CardinalMarkE',
                  22: 'CardinalMarkS',
                  23: 'CardinalMarkW',
                  24: 'PortHandMark',
                  25: 'StarboardHandMark',
                  26: 'PreferredChannelPortHand',
                  27: 'PreferredChannelStarboardHand',
                  28: 'IsolatedDanger',
                  29: 'SafeWater',
                  30: 'SpecialMark',
                  31: 'LightVessel/LANBY/Rigs'
                }

 }

File: client/dart-client/Vessel.dart
library Vessel;

import 'LeafletMap.dart';
import 'MapFeature.dart';
import 'ais-html5.dart';
import 'dart:math';

class Vessel{
  var pos;
  num mmsi, msgid, ship_type, nav_status;
  String name, imo, dest;
  num cog, sog, true_heading, dim_port, dim_stern, dim_bow, dim_starboard, draught, brng;
  int time_received, time_captured;
  MapFeature vector, polygon, feature;
  const MINIMUM_SPEED = 0.4;
  
  Vessel(jsonObject){
    mmsi = (jsonObject['userid'] !=null) ? jsonObject['userid']:jsonObject['mmsi']; //notwendig, weil posEvents userid statt mmsi haben
    msgid = jsonObject['msgid'];
    name = jsonObject['name'];
    time_received = jsonObject['time_received'];
    cog = jsonObject['cog'];
    sog = jsonObject['sog'];
    pos = jsonObject['pos'];
    imo = jsonObject['imo'];
    true_heading = jsonObject['true_heading'];
    dim_port = jsonObject['dim_port'];
    dim_stern = jsonObject['dim_stern'];
    dim_bow = jsonObject['dim_bow'];
    dim_starboard = jsonObject['dim_starboard'];
    ship_type = jsonObject['ship_type'];
    nav_status = jsonObject['nav_status'];
    dest = jsonObject['dest'];
    draught = jsonObject['draught'];
    time_captured = jsonObject['time_captured'];
    if (mmsi == 211855000) //Cap San Diego
    {
      true_heading = 299;
    }
  }
  updatePosition(json){
    msgid =  json['msgid'];
    time_received = json['time_received'];
    cog = json['cog'];
    sog = json['sog'];
    pos = json['pos'];
    true_heading = json['true_heading'];
  }
  
  paintToMap(zoom, callback){
    if(pos != null)
    {
        var moving = (sog !=null && sog >= MINIMUM_SPEED && sog!=102.3) ; 
        var shipStatics = (zoom > 11) &&  (cog !=null ||(true_heading!=null && true_heading!=0.0 && true_heading !=511)) && (dim_port !=null && dim_stern!=null) ;

        brng = calcAngle(sog, cog, true_heading);
        var cos_angle = cos(brng);
        var sin_angle = sin(brng);
        List<Coord> vectorPoints = new List();
        Coord shipPoint = new Coord(pos[1],pos[0]);
        vectorPoints.add(shipPoint);

        if (moving) //only vessels, that move with a minimum speed of MINIMUM_SPEED
        {
          var meterProSekunde = sog *0.51444;
          var vectorLength = meterProSekunde * 30; //meters, which are covered in 30 sec
          var targetPoint = destinationPoint(pos, cog, vectorLength);
          vectorPoints.add(targetPoint);
          var vectorWidth = (sog > 30?5:2); 
          vector = new Polyline(vectorPoints, {'color': 'red', 'weight': vectorWidth });
          vector.addToMap(false, "");
          var animationPartsSize = vectorLength/(zoom*10); //how long are the chunks of the vector
          var animationInterval = 400; //how long is the interval between two animation steps
          if (shipStatics)
          {
            polygon = new AnimatedPolygon(vectorPoints,{
              'autoStart':false,
              'distance': animationPartsSize,
              'interval': animationInterval,
              'dim_stern':dim_stern,
              'dim_port': dim_port,
              'dim_bow':dim_bow,
              'dim_starboard': dim_starboard,
              'brng':brng,
              'color': "blue",
              'weight': 3,
              'fill':true,
              'fillColor':shipTypeColors[ship_type],
              'fillOpacity':0.6,
              'clickable':false
            });
          }
          feature = new AnimatedPolygon(vectorPoints,{
            'autoStart': false,
            'distance': animationPartsSize,
            'interval':animationInterval,
            'brng':brng,
            'zoom': LMap.getZoom(),
            'color': "black",
            'weight': 1,
            'fill':true,
            'fillColor': shipTypeColors[ship_type],
            'fillOpacity':0.8,
            'clickable':true
             });

        }/* paint for non moving vessels a Polygon and a circlemarker*/
        else 
        {
          if(shipStatics)
          {
            polygon = new AnimatedPolygon(vectorPoints,{
                'autoStart':false,
                'dim_stern':dim_stern,
                'dim_port': dim_port,
                'dim_bow':dim_bow,
                'dim_starboard': dim_starboard,
                'brng':brng,
                'color': "blue",
                'weight': 3,
                'fill':true,
                'fillColor':shipTypeColors[ship_type],
                'fillOpacity':0.6,
                'clickable':false,
                'zoom':false
              });
          }
          feature = new CircleMarker(vectorPoints[0], {
                           'radius':5,
                           'fill':true,
                           'fillColor':shipTypeColors[ship_type],
                           'fillOpacity':0.8,
                           'color':"#000000",
                           'opacity':0.4,
                           'weight':2.5 
                           });
        }
        if(polygon != null)
        {
          polygon.addToMap(moving, ""); 
        }
        feature.addToMap(moving, createPopupContent());
    }
  callback();
}
  
  String createPopupContent(){
    String mouseOverPopup ="<div class='mouseOverPopup'><table>";
    if(name != null) {
        mouseOverPopup = "${mouseOverPopup}<tr><td colspan='2'><b>${name}</b></nobr></td></tr>";
      }
      if(imo != null && imo !="0") {
        mouseOverPopup = "${mouseOverPopup}<tr><td>IMO</td><td>${imo}</b></nobr></td></tr>";
      }
      mouseOverPopup = "${mouseOverPopup}<tr><td>MMSI: &nbsp;</td><td><nobr>${mmsi.toString()}</nobr></td></tr>";
    if(nav_status!=null && nav_status < 15 && nav_status > -1) {
      mouseOverPopup = "${mouseOverPopup}<tr><td>NavStatus: &nbsp;</td><td><nobr>${nav_stati[nav_status]}</nobr></td></tr>";
    }
    if(sog!= null) {
      mouseOverPopup = "${mouseOverPopup}<tr><td>Speed: &nbsp;</td><td><nobr>${sog.toString()} kn</nobr></td></tr>";
    }
    if(true_heading != null && true_heading != 511) {
      mouseOverPopup = "${mouseOverPopup}<tr><td>Heading: &nbsp;</td><td><nobr>${true_heading.toString()} °</nobr></td></tr>";
    }
    else if(cog != null) {
      mouseOverPopup = "${mouseOverPopup}<tr><td>Course: &nbsp;</td><td><nobr>${cog.toString()} °</nobr></td></tr>";
    }
    mouseOverPopup = "${mouseOverPopup}<tr><td>TimeReceived: &nbsp;</td><td><nobr>${getDate(time_received)}</nobr></td></tr>";
    if(dest != null) {
      mouseOverPopup = "${mouseOverPopup}<tr><td>Dest</td><td>${dest}</b></nobr></td></tr>";
    }
    if(draught != null) {
      mouseOverPopup = "${mouseOverPopup}<tr><td>draught</td><td>${(draught/10).toString()} m</b></nobr></td></tr>";
    }
    if(dim_bow != null && dim_port != null) {
      mouseOverPopup = "${mouseOverPopup}<tr><td>length</td><td>${(dim_stern+dim_bow).toString()} m</b></nobr></td></tr>";
    }
    if (shipTypes[ship_type] != null) {
      mouseOverPopup = "${mouseOverPopup}<tr><td>ship_type</td><td>${shipTypes[ship_type]}</b></nobr></td></tr>";
    }
    mouseOverPopup = "${mouseOverPopup}</table></div>";
    return mouseOverPopup;
  }
}

const EARTH_RADIUS = 6371000; 

String getDate(timeReceived){
  var timeString = (new DateTime.fromMillisecondsSinceEpoch(timeReceived,isUtc: false)).toString();
  var month = timeString.substring(5, 7);
  var day = timeString.substring(8,10);
  var time = timeString.substring(10,16);
  return "${day}.${month}. ${time}";
}

num calcAngle(sog, cog, true_heading) {
    var direction = 0;
    if (sog!=null && sog > 0.4 && cog < 360)
    {
      direction = cog;
    }
    else if (true_heading!=null && true_heading > 0.0 && true_heading !=511 && true_heading < 360)
    {
      direction = true_heading;
    }
    return (-direction *(PI / 180.0));
  }

  Coord destinationPoint(pos, cog, dist) {
    dist = dist / EARTH_RADIUS;  
    var brng = cog * (PI / 180);  
    var lat = pos[1] * (PI / 180);
    var lon = pos[0] * (PI / 180);
    var lat_dest = asin(sin(lat) * cos(dist) + cos(lat) * sin(dist) * cos(brng));
    var lon_dest = lon + atan2(sin(brng) * sin(dist) * cos(lat), cos(dist) - sin(lat) * sin(lat_dest));
    if (lat_dest.isNaN || lon_dest.isNaN) return null;
    lat_dest = lat_dest *(180/PI);
    lon_dest = lon_dest *(180/PI);
    return new Coord(lat_dest, lon_dest);
  }
  Map<int, String> shipTypes = new Map<int, String>();
  Map<int,String> shipTypeColors = new Map<int,String>();
  Map<int,String> nav_stati = new Map<int,String>();

  initTypeMaps(){
  
  shipTypes[2] = 'Other Type';
  shipTypes[6] = 'Passenger Ships';
  shipTypes[7] = 'Cargo Ships';
  shipTypes[8] = 'Tankers';
  shipTypes[9] = 'Other Type';
  shipTypes[20] = 'Wing in ground (WIG)';
  shipTypes[29] = 'Wing in ground (WIG)';
  shipTypes[30] ='Fishing';
  shipTypes[31] ='Towing';
  shipTypes[32] ='Towing';
  shipTypes[33] ='Dredger';
  shipTypes[34] ='diving operations';
  shipTypes[35] ='military operations';
  shipTypes[36] = 'Sailing';
  shipTypes[37] = 'Pleasure craft';
  shipTypes[38] = 'Reserved';
  shipTypes[39] = 'Reserved';
  shipTypes[40] = 'High speed craft';
  shipTypes[49] = 'High speed craft';
  shipTypes[50] ='Pilot vessel';
  shipTypes[51] ='Search and rescue vessels';
  shipTypes[52] ='Tugs';
  shipTypes[53] ='Port tenders';
  shipTypes[54] ='anti-pollution vessels';
  shipTypes[55] ='Law enforcement vessels';
  shipTypes[56] ='Spare for local vessels';
  shipTypes[57] ='Spare for local vessels';
  shipTypes[58] ='Medical transports';
  shipTypes[59] = 'Ships according to RR';
  shipTypes[60] = 'Passenger Ships';
  shipTypes[61] = 'Passenger Ships';
  shipTypes[63] = 'Passenger Ships';
  shipTypes[65] = 'Passenger Ships';
  shipTypes[67] = 'Passenger Ships';
  shipTypes[69] = 'Passenger Ships';
  shipTypes[70] = 'Cargo Ships';
  shipTypes[71] = 'Cargo Ships';
  shipTypes[72] = 'Cargo Ships';
  shipTypes[73] = 'Cargo Ships';
  shipTypes[74] = 'Cargo Ships';
  shipTypes[77] = 'Cargo Ships';
  shipTypes[79] = 'Cargo Ships';
  shipTypes[80] = 'Tanker';
  shipTypes[81] = 'Tanker';
  shipTypes[82] = 'Tanker';
  shipTypes[83] = 'Tanker';
  shipTypes[84] = 'Tanker';
  shipTypes[89] = 'Tanker';
  shipTypes[90] = 'Other Type';
  shipTypes[91] = 'Other Type';
  shipTypes[97] = 'Other Type';
  shipTypes[99] = 'Other Type';
  
  
 
  shipTypeColors[2] ='#f9f9f9';
  shipTypeColors[20] ='#f9f9f9';
  shipTypeColors[29] ='#f9f9f9';
  shipTypeColors[30] ='#f99d7b'/*brown, Fishing*/;
  shipTypeColors[31] ='#4dfffe'/*lightblue, Towing*/;
  shipTypeColors[32] ='#4dfffe'/*lightblue, Towing*/;
  shipTypeColors[33] ='#f9f9f9'/*gray, Dredger*/;
  shipTypeColors[34] ='white'/*Engaged in diving operations*/;
  shipTypeColors[35] ='white'/*Engaged in military operations*/;
  shipTypeColors[36] ='#f900fe'/*violett, Sailing*/;
  shipTypeColors[37] ='#f900fe'/*violett, Pleasure craft*/;
  shipTypeColors[40] ='#f9f9f9'/*Highspeed*/;
  shipTypeColors[49] ='#f9f9f9'/*Highspeed*/;
  shipTypeColors[50] ='red'/*Pilot vessel*/;
  shipTypeColors[51] ='white' /*Search and rescue vessels*/;
  shipTypeColors[52] ='#4dfffe'/*lightblue, Tugs*/;
  shipTypeColors[53] ='#4dfffe'/*lightblue, Port tenders*/;
  shipTypeColors[54] ='white'/*anti-pollution vessels*/;
  shipTypeColors[55] ='white'/*Law enforcement vessels*/;
  shipTypeColors[56] ='#d2d2d2'/*not classified => used as default by vesseltracker*/;
  shipTypeColors[57] ='white'/*Spare for local vessels*/;
  shipTypeColors[58] ='white'/*Medical transports*/;
  shipTypeColors[59] ='white'/*Ships according to RR*/;
  shipTypeColors[6] ='#2d00fe'/*blue, Passenger Ships*/;
  shipTypeColors[60] ='#2d00fe'/*blue, Passenger Ships*/;
  shipTypeColors[61] ='#2d00fe'/*blue, Passenger Ships*/;
  shipTypeColors[63] ='#2d00fe'/*blue, Passenger Ships*/;
  shipTypeColors[65] ='#2d00fe'/*blue, Passenger Ships*/;
  shipTypeColors[67] ='#2d00fe'/*blue, Passenger Ships*/;
  shipTypeColors[69] ='#2d00fe'/*blue, Passenger Ships*/;
  shipTypeColors[7] = '#95f190'/*lightgreen, Cargo Ships*/;
  shipTypeColors[70] ='#95f190'/*lightgreen, Cargo Ships*/;
  shipTypeColors[71] ='#95f190'/*lightgreen, Cargo Ships*/;
  shipTypeColors[72] ='#95f190'/*lightgreen, Cargo Ships*/;
  shipTypeColors[73] ='#95f190'/*lightgreen, Cargo Ships*/;
  shipTypeColors[74] ='#95f190'/*lightgreen, Cargo Ships*/;
  shipTypeColors[77] ='#95f190'/*lightgreen, Cargo Ships*/;
  shipTypeColors[79] ='#95f190'/*lightgreen, Cargo Ships*/;
  shipTypeColors[8] = '#f70016'/*red, Tankers*/;
  shipTypeColors[80] ='#f70016'/*Tanker*/;
  shipTypeColors[81] = '#f70016'/*red, Tankers*/;
  shipTypeColors[82] = '#f70016'/*red, Tankers*/;
  shipTypeColors[83] = '#f70016'/*red, Tankers*/;
  shipTypeColors[84] = '#f70016'/*red, Tankers*/;
  shipTypeColors[89] ='#f70016'/*red,Tankers*/;
  shipTypeColors[9] ='#d2d2d2'/*Other Type*/;
  shipTypeColors[90] ='#d2d2d2'/*Other Type*/;
  shipTypeColors[91] ='#d2d2d2'/*Other Type*/;
  shipTypeColors[97] ='#d2d2d2'/*Other Type*/;
  shipTypeColors[99] ='#d2d2d2'/*Other Type*/;

  
  nav_stati[0] ='under way us. engine';
  nav_stati[1] ='at anchor';
  nav_stati[2] = 'not under command';
  nav_stati[3] = 'restr. maneuverability';
  nav_stati[4] = 'constr. by draught';
  nav_stati[5] = 'moored';
  nav_stati[6] = 'aground';
  nav_stati[7] = 'engaged in fishing';
  nav_stati[8] = 'under way sailing';
  nav_stati[9] = 'future use';
  nav_stati[10] = 'future use';
  nav_stati[11] = 'future use';
  nav_stati[12] = 'future use';
  nav_stati[13] = 'future use';
  nav_stati[14] = 'AIS-SART (active)';
  nav_stati[15] = 'not defined';
}